pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -1414,6 +1464,208 @@ func compareCAPIInfraMachines(platform configv1.PlatformType, infraMachine1, inf
+	// MAPI Machine does not have the observed generation field.
+
+	isPatchRequired, err := util.IsPatchRequired(existingMAPIMachine, patchBase)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if patch is required: %w", err)
+	}
+
+	if !isPatchRequired {
+		// If the patch is not required, return early.
+		return false, nil
+	}
+
+	if err := r.Status().Patch(ctx, existingMAPIMachine, patchBase); err != nil {
+		logger.Error(err, "Failed to update Machine API machine status")
+		updateErr := fmt.Errorf("failed to update status: %w", err)
+
+		if condErr := r.applySynchronizedConditionWithPatch(
+			ctx, existingMAPIMachine, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachine, updateErr.Error(), nil); condErr != nil {
+			return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+		}
+
+		return false, updateErr
+	}
+
+	return true, nil
+}
+
+// ensureMAPIMachineSpecUpdated updates the MAPI machine if changes are detected.
+func (r *MachineSyncReconciler) ensureMAPIMachineSpecUpdated(ctx context.Context, existingMAPIMachine *mapiv1beta1.Machine, mapiMachinesDiff map[string]any, updatedMAPIMachine *mapiv1beta1.Machine) (bool, error) {
+	logger := logf.FromContext(ctx)
+
+	// If there are no spec changes, return early.
+	if !hasSpecOrMetadataOrProviderSpecChanges(mapiMachinesDiff) {
+		return false, nil
+	}
+
+	logger.Info("Changes detected for Machine API machine. Updating it", "diff", fmt.Sprintf("%+v", mapiMachinesDiff))
+
+	if err := r.Update(ctx, updatedMAPIMachine); err != nil {
+		logger.Error(err, "Failed to update Machine API machine")
+
+		updateErr := fmt.Errorf("failed to update Machine API machine: %w", err)
+
+		if condErr := r.applySynchronizedConditionWithPatch(ctx, existingMAPIMachine, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachine, updateErr.Error(), nil); condErr != nil {
+			return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+		}
+
+		return false, updateErr
+	}
+
+	return true, nil
+}
+
+// setChangedCAPIMachineStatusFields sets the updated fields in the CAPI machine status.
+func setChangedCAPIMachineStatusFields(existingCAPIMachine, convertedCAPIMachine *clusterv1.Machine) {
+	// convertedCAPIMachine holds the computed and desired status changes converted from the source MAPI machine, so apply them to the existing existingCAPIMachine.
+	// Merge the v1beta1 conditions.
+	for i := range convertedCAPIMachine.Status.Conditions {
+		conditions.Set(existingCAPIMachine, &convertedCAPIMachine.Status.Conditions[i])
+	}
+
+	// Copy them back to the convertedCAPIMachine.
+	convertedCAPIMachine.Status.Conditions = existingCAPIMachine.Status.Conditions
+
+	// Merge the v1beta2 conditions.
+	if convertedCAPIMachine.Status.V1Beta2 != nil && existingCAPIMachine.Status.V1Beta2 != nil {
+		for i := range convertedCAPIMachine.Status.V1Beta2.Conditions {
+			conditionsv1beta2.Set(existingCAPIMachine, convertedCAPIMachine.Status.V1Beta2.Conditions[i])
+		}
```

Comment:
Sorry but we have to iterate again.

We need to ensure to always use `conditionsv1beta2.Set`, because otherwise LastTransitionTime for v1beta2 conditions might not be set. `LastTransitionTime` is a required field with v1.11 though.

```suggestion
	if convertedCAPIMachine.Status.V1Beta2 != nil {
        if existingCAPIMachine.Status.V1Beta2 == nil {
			existingCAPIMachine.Status.V1Beta2 = &clusterv1.MachineV1Beta2Status{}
		}

		for i := range convertedCAPIMachine.Status.V1Beta2.Conditions {
			conditionsv1beta2.Set(existingCAPIMachine, convertedCAPIMachine.Status.V1Beta2.Conditions[i])
		}
```

Reply:
Ah good point! Missed this :)